src/exec/exec.go

```diff
@@ -0,0 +1,78 @@
+package exec
+
+import (
+	"context"
+	"math"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"gopkg.in/op/go-logging.v1"
+
+	"github.com/thought-machine/please/src/core"
+	"github.com/thought-machine/please/src/fs"
+	"github.com/thought-machine/please/src/process"
+)
+
+var log = logging.MustGetLogger("exec")
+
+func Exec(state *core.BuildState, label core.BuildLabel, cmdArgs []string, shareNetwork bool) {
+	exec(context.Background(), state, label, cmdArgs, shareNetwork)
+}
+
+func exec(ctx context.Context, state *core.BuildState, label core.BuildLabel, cmdArgs []string, shareNetwork bool) error {
+	target := state.Graph.TargetOrDie(label)
+
+	if !target.IsBinary && len(cmdArgs) == 0 {
+		log.Fatalf("Target %s cannot be executed; it's not marked as binary", target.Label)
+	}
+
+	var cmd string
+	if len(cmdArgs) > 0 {
+		var err error
+		if cmd, err = core.ReplaceSequences(state, target, strings.Join(cmdArgs, " ")); err != nil {
+			return err
+		}
+	} else {
+		outs := target.Outputs()
+		if len(outs) != 1 {
+			log.Fatalf("Target %s cannot be executed as it has %d outputs", target.Label, len(outs))
+		}
+		cmd = outs[0]
```

Comment:
I think `plz run` does something slightly different where it uses `$(out_exe ...)`. Not sure if there's a good reason for it though.

Reply:
Yeah, I need to come back to this